atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKvsAsyncFallbackMechanismsTests.java

```diff
@@ -0,0 +1,120 @@
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.keyvalue.cassandra.async.CassandraAsyncKeyValueServiceFactory;
+
+public class CassandraKvsAsyncFallbackMechanismsTests {
+    private static final TableReference TEST_TABLE = TableReference.createFromFullyQualifiedName("ns.pt_kvs_test");
+    private static final Cell CELL = Cell.create(PtBytes.toBytes("row"), PtBytes.toBytes("column"));
+
+    @ClassRule
+    public static final CassandraResource CASSANDRA_RESOURCE = new CassandraResource();
+
+    private KeyValueService keyValueService;
+    private Proxy socksProxy;
+    private AsyncKeyValueService asyncKeyValueService;
+    private CassandraAsyncKeyValueServiceFactory factory;
+
+    private static ImmutableCassandraKeyValueServiceConfig getConfigWithGcGraceSeconds(
+            CassandraKeyValueServiceConfig config,
+            int gcGraceSeconds) {
+        return ImmutableCassandraKeyValueServiceConfig
+                .copyOf(config)
+                .withGcGraceSeconds(gcGraceSeconds);
+    }
+
+
+    @Before
+    public void setUp() {
+        asyncKeyValueService = mock(AsyncKeyValueService.class);
+        when(asyncKeyValueService.getAsync(any(), any())).thenReturn(Futures.immediateFuture(ImmutableMap.of()));
+        factory = mock(CassandraAsyncKeyValueServiceFactory.class);
+    }
+
+    @After
+    public void tearDown() {
+        try {
+            keyValueService.truncateTables(ImmutableSet.of(TEST_TABLE));
+        } catch (Exception e) {
+            // this is fine
+        }
+        CASSANDRA_RESOURCE.registerKvs(keyValueService);
+    }
+
+    @Test
+    public void testGetFallBackNotNeeded() {
+        when(factory.constructAsyncKeyValueService(any(), any(), eq(AtlasDbConstants.DEFAULT_INITIALIZE_ASYNC)))
+                .thenReturn(Optional.of(asyncKeyValueService));
+
+        CassandraKeyValueServiceConfig config = ImmutableCassandraKeyValueServiceConfig.builder()
+                .from(CASSANDRA_RESOURCE.getConfig())
+                .asyncKeyValueServiceFactory(factory)
+                .build();
+
+        keyValueService = CassandraKeyValueServiceImpl.createForTesting(config);
+        keyValueService.getAsync(TEST_TABLE, ImmutableMap.of(CELL, 3L));
+
+        verify(asyncKeyValueService, times(1)).getAsync(any(), any());
+    }
+
+    @Test
+    public void testGetFallingBackToSynchronous() {
+        when(factory.constructAsyncKeyValueService(any(), any(), eq(AtlasDbConstants.DEFAULT_INITIALIZE_ASYNC)))
+                .thenReturn(Optional.empty());
+
+        CassandraKeyValueServiceConfig config = ImmutableCassandraKeyValueServiceConfig.builder()
+                .from(CASSANDRA_RESOURCE.getConfig())
+                .asyncKeyValueServiceFactory(factory)
+                .build();
+
+        keyValueService = CassandraKeyValueServiceImpl.createForTesting(config);
+        keyValueService.createTable(TEST_TABLE, AtlasDbConstants.GENERIC_TABLE_METADATA);
```

Comment:
non-actionable: This seems a bit strange here, though I gather that you need it for the test to pass since it _actually_ hits the sync endpoint.

Reply:
It didn't make sense to throw so I did it like this.